pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,338 @@
+limitations under the License.
+*/
+
+package container
+
+import (
+	"io"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+type PodSandboxID string
+
+// PodSandboxManager provides basic operations to create/delete and examine the
+// PodSandboxes.
+type PodSandboxManager interface {
+	// Create creates a sandbox based on the given config, and returns the
+	//  of the new sandbox.
+	Create(config *PodSandboxConfig) (PodSandboxID, error)
+	// Delete deletes the sandbox by its ID. If there are any running
+	// containers in the sandbox, they will be terminated as a side-effect.
+	Delete(id PodSandboxID) error
+	// List lists existing sandboxes, filtered by the given PodSandboxFilter.
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	// Status gets the status of the sandbox by ID.
+	Status(id PodSandboxID) (PodSandboxStatus, error)
+	// PortForward copies the data from the port in the PodSandBox to the
+	// stream.
+	// TODO: Should PortForward be in a separate interface.
+	PortForward(id PodSandboxID, port uint16, stream io.ReadWriteCloser) error
+}
+
+// PodSandboxConfig holds all the required and optional fields for creating a
+// sandbox.
+type PodSandboxConfig struct {
+	// Name is the name of the sandbox.
+	Name string
+	// Hostname is the hostname of the sandbox.
+	Hostname string
+	// DNSOptions sets the DNS options for the sandbox.
+	DNSOptions DNSOptions
+	// PortMappings lists the port mappings for the sandbox.
+	PortMappings []PortMapping
+	// Mounts lists the mounts to be added to the sandbox's filesystem.
+	Mounts []Mount
+	// Resources specifies the resource requirements for the sandbox.
+	// Note: On a Linux host, kubelet will create a pod-level cgroup and pass
+	// it as the cgroup parent for the PodSandbox. For some runtimes, this is
+	// sufficent to fulfill pod-level resource requirements, and this field
+	// will not be used. For others, e.g., hypervisor-based runtimes, explicit
+	// resource requirements for the sandbox are needed at creation time.
+	Resources PodSandboxResources
+	// Labels are key value pairs that may be used to scope and select individual resources.
+	Labels Labels
+	// Annotations is an unstructured key value map that may be set by external
+	// tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string
+
+	// Linux contains configurations specific to Linux hosts.
+	Linux *LinuxPodSandboxConfig
+}
+
+// Labels are key value pairs that may be used to scope and select individual resources.
+// Label keys are of the form:
+//     label-key ::= prefixed-name | name
+//     prefixed-name ::= prefix '/' name
+//     prefix ::= DNS_SUBDOMAIN
+//     name ::= DNS_LABEL
+type Labels map[string]string
+
+// LinuxPodSandboxConfig holds platform-specific configuraions for Linux
+// host platforms and Linux-based containers.
+type LinuxPodSandboxConfig struct {
+	// CgroupParent is the parent cgroup of the sandbox.
+	CgroupParent string
```

Comment:
@vishh - this is complicated, and I want to clarify a few things.

1. ability to manage usage per qos-tier (not covered in this spec, but stating for awareness)
1. ability to manage usage per pod
1. ability to manage usage per container

The first thing we should clarify is who creates the `cgroups` associated with 2 and 3.

* Does the pod sandbox manager create both?  
* Or does it just create 2 and it delegates to the runtime to create 3 as a child?

Let's assume its the latter, and evaluate two runtime options: `docker` and `runc`.

If `kubelet` ran with `--cgroup-driver=systemd`, and user creates `podA`:

* pod sandbox manager creates a `podA.slice`
* instructs container runtime to B and C
 * docker with `--exec-opt native.cgroupdriver=systemd`, the runtime *requires* systemd style cgroup name
 * docker with `--exec-opt native.cgroupdriver=cgroupfs`, the runtime will expect a cgroupfs style cgroup name, if we wanted to pass a cgroupfs syntax of the slice, then `podA.slice` would have had to set `Delegate=yes`.
 * `runc --systemd-cgroup start` would need to covert the name `podA.slice:runc:containerId`
 * `runc` without that global option would need `Delegate=yes`

In the `runc` case, I suspect we would always pass the `--systemd-cgroup` option if the `kubelet` was launched with `--cgroup-driver=systemd`.

I suspect with `rkt` they will want the `slice` name in `systemd` format to launch but maybe @jonboulle can weigh in.

Summary:
How the kubelet creates the slice is dependent upon the container runtime configuration for how to set `Delegate`.

@mrunalp - fyi per discussion on having a way to introspect cgroup driver for docker so we can know the expected input format.